Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,243 @@
+        /// Abstracted value for the how long the dwelled object still needs to be focused to comlete the dwell action
+        /// </summary>
+        protected float FillTimer = 0;
+
+        private void Awake()
+        {
+            Debug.Assert(dwellProfile != null, "DwellProfile is null, creating default profile.");
+
+            if (dwellProfile == null)
+            {
+                dwellProfile = ScriptableObject.CreateInstance<DwellProfile>();
+            }
+
+            CancelDwellAction += OnDwellCancelAction;
+        }
+
+        private void Update()
+        {
+            UpdateFillTimer();
+
+            if (HasFocus && CurrentDwellState != DwellStateType.DwellCompleted)
+            {
+                double focusDuration = (DateTime.UtcNow - this.focusEnterTime).TotalSeconds;
+
+                if (CurrentDwellState == DwellStateType.FocusGained && focusDuration >= dwellProfile.DwellIntentDelay)
+                {
+                    DwellIntended.Invoke(pointer);
+                    CurrentDwellState = DwellStateType.DwellIntended;
+                }
+                else if (CurrentDwellState == DwellStateType.DwellIntended && (focusDuration - dwellProfile.DwellIntentDelay) >= dwellProfile.DwellStartDelay)
+                {
+                    DwellStarted.Invoke(pointer);
+                    CurrentDwellState = DwellStateType.DwellStarted;
+                    FillTimer = dwellProfile.TimeToCompleteDwell;
+                }
+                else if (CurrentDwellState == DwellStateType.DwellStarted && (focusDuration - dwellProfile.DwellIntentDelay - dwellProfile.DwellStartDelay) >= dwellProfile.TimeToCompleteDwell)
+                {
+                    DwellCompleted.Invoke(pointer);
+                    CurrentDwellState = DwellStateType.DwellCompleted;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Default FillTimer computation based on profile settings
+        /// </summary>
+        protected virtual void UpdateFillTimer()
+        {
+            switch (CurrentDwellState)
+            {
+                case DwellStateType.None:
+                case DwellStateType.FocusGained:
+                case DwellStateType.DwellIntended:
+                    FillTimer = 0;
+                    break;
+                case DwellStateType.DwellStarted:
+                    FillTimer -= Time.deltaTime;
+                    FillTimer = FillTimer < 0 ? 0 : FillTimer;
+                    break;
+                case DwellStateType.DwellCompleted:
+                    Debug.Assert(FillTimer < float.Epsilon, $"Fill timer was not 0. It was {FillTimer}.");
+                    break;
+                case DwellStateType.DwellCanceled:
+                    if ((DateTime.UtcNow - focusExitTime).TotalSeconds > dwellProfile.TimeToAllowDwellResume)
+                    {
+                        FillTimer = 0;
+                        CurrentDwellState = DwellStateType.None;
```

Comment:
>CurrentDwellState = DwellStateType.None; [](start = 24, length = 40)

It feels weird that a function called UpdateFillTimer will also update the current dwell state - also interesting because subclasses could forget to update this dwell state change.

If this is intended to address a specific issue, it might be worth having a comment about what this is for

Reply:
Yeah it is ok for the deriving class to "forget". In most dwell variations I have seen the DwellCanceled state can transition to None or Started based on conditions. Those conditions can vary based on the profile supplied to the implementation. It can be left up to the developer to decide what state transition they want.